Switch to use uPheno base and import phenotype ontologies separately

matentzn opened the pull request:
Fixes #24 #34 #36 #37 #82 #88 #91 #96 

But really is a refactoring of the full import system.

This ensures that we always have the latest phenotype ontologies in phenio, independent of any particular phenio release. I have eyeballed the resulting ontology and it has still waaaay too many dangling classes, I will try to clean those in subsequent commits.

@caufieldjh I made this a draft because I need to do quite a bit of clean up of phenio which has become quite messy in terms of dangling classes. Will come back to you updates, but feel free to look at the PR and tell me your general opinion of my suggestion

matentzn committed:
Switch to use uPheno base and import phenotype ontologies separately

This ensures that we always have the latest phenotype ontologies in phenio, independent of any particular phenio release. I have eyeballed the resulting ontology and it has still waaaay too many dangling classes, I will try to clean those in subsequent commits

caufieldjh commented:
makes sense to me conceptually - seems like it could become an issue if there's discrepancies between the most recent version of a phenotype ontology and upheno, but I will trust you on this

matentzn committed:
Update phenio-odk.yaml

matentzn commented on src/ontology/phenio-odk.yaml:
@caufieldjh @kevinschaper can you review just this file please while I am cleaning stuff? Ever line matters; if you don't understand a change better ask. Big implications for everything.

caufieldjh replied:
Is there a bridge file for upheno to OBA already?

caufieldjh replied:
Some of these newly imported ID resources like CGNC are new to me - are they coming in because Upheno uses them?

matentzn replied:
https://github.com/monarch-initiative/phenio/pull/94/files#r1848735596

matentzn commented on src/ontology/phenio-odk.yaml:
```diff
@@ -18,20 +18,67 @@ import_group:
+    - <http://identifiers.org/hgnc/*>
+    - <http://identifiers.org/ncbigene/*>
+    # From PRO:
+    - <http://www.ncbi.nlm.nih.gov/gene/*>
```
@caufieldjh all the above, thats why this list is important, are _excluded_ ID spaces. If you feel something is not rightfully _excluded_, we need to properly import it.

caufieldjh replied:
ah! I see. In that case I will also suggest excluding:
* <http://purl.obolibrary.org/obo/STATO_*>
* DOID
* Wikidata prefixes (WD_Entity and WD_Prop)

caufieldjh replied:
otherwise I'm happy with the list barring any objection from @kevinschaper

matentzn committed:
Update phenio-odk.yaml

matentzn committed:
Removing unnecessary namespaces from final phenio.owl

matentzn committed:
Stripping the edit file of all things but metadata

matentzn committed:
Add CL and Uberon mappings

matentzn committed:
Remove import from git index

It has become too big

matentzn committed:
Update ODK config and related files

matentzn committed:
Create file for ROBOT renames

matentzn committed:
Added MAXO component

matentzn committed:
Add MAXO to ODK config

matentzn committed:
MISC refactor phenio.Makefile

1. added custom goals for go and emapa to enable additional processing steps (in particular creating an artificial root term for emapa so the ontology looks better in a browser)
2. Remove the biolink node assignment entirely from phenio
3. Exclude many more namespaces
4. Removing all the uberon bridges in favour of amazingly simple approach using sssom:inject (with uberon and cl mappings).

matentzn commented on src/ontology/config/mappings-to-uberon-bridge.rules:
@gouttegd would you mind reviewing this rules file here? It seems to work:

<img width="1094" alt="image" src="https://github.com/user-attachments/assets/cf7581d7-417b-401b-8c6c-d3ed4dd30d4e">


But maybe I am doing something stupid. 

I am missing one thing: 

I would like

```
subject==UBERON:* -> invert();
!object==UBERON:* -> stop();
```

to be

```
subject==(UBERON|CL):* -> invert();
!object==(UBERON|CL):* -> stop();
```

but that didn't work.. What is the proper way to achieve this?

BTW I solved this problem by a mix of resorting to the docs and looking at Uberon pipeline. So the only feedback on the docs is from my perspective to just provide two or three more examples.

gouttegd replied:
> ```
> subject==(UBERON|CL):* -> invert();
> ```

Filters don’t work that way. Every atomic filter (such as `subject`) can only filter on one value. To select mappings with a subject that is either in Uberon or in CL, you need to compose two distinct `subject` filters:

```
(subject==UBERON:* || subject==CL:*) -> invert();
!(object==UBERON:* || object==CL:*) -> stop();
```

Of note, in the first rule (`invert`), the parentheses around the two atomic filters are not strictly necessary (the rule would have the same meaning without them), but I do think they make the intention clearer. In the second rule however (`stop`), the parentheses _are_ strictly necessary, because

```
!object==UBERON:* || object==CL:* -> stop();
```

would be interpreted as: “drop all mappings whose object is not in Uberon or whose object is in CL”.

matentzn replied:
Thaaaaank you!!! Exciting!

gouttegd commented on src/ontology/config/mappings-to-uberon-bridge.rules:
```diff
@@ -0,0 +1,16 @@
+!object==UBERON:* -> stop();
+
+# Ignore any mapping to/from an inexistent or obsolete class.
+predicate==* -> check_object_existence();
```
Just for information, the next version of SSSOM-Java will include a newer and (I believe) cleaner syntax to do this — but the existing syntax will remain supported, so nothing to worry about.

matentzn replied:
I will watch out for that!! Thanks!

gouttegd replied:
Basically there will be a new filter called `exists()`, allowing to perform an arbitrary action based on whether an entity exists (and is not deprecated) or not in the ontology.

The equivalent of `check_object_existence()` will be:

```
!exists(%{object_id}) -> stop();
```

(Selects all mappings whose object does _not_ — because the filter is negated, `!` – exist in the ontology, and drop them.)

matentzn replied:
Nice! Thanks for letting me know in advance!

gouttegd commented on src/ontology/config/mappings-to-uberon-bridge.rules:
```diff
@@ -0,0 +1,16 @@
+# Ignore any mapping where the same foreign term is mapped to more than one UBERON class.
+!cardinality==*:1 -> stop();
+
+# Create the hasDbXref annotation and annotate it with a subset of the mapping metadata
```
I just note that this comment does not match what the rule does. The rule creates a `SubClassOf` axiom, it has nothing to do with a `oboInOwl:hasDbXref` annotation.

matentzn replied:
Copy paste error, fixed now thank you!

gouttegd commented on src/ontology/config/mappings-to-uberon-bridge.rules:
```diff
@@ -0,0 +1,16 @@
+!cardinality==*:1 -> stop();
+
+# Create the hasDbXref annotation and annotate it with a subset of the mapping metadata
+object==UBERON:* && predicate==semapv:crossSpeciesExactMatch -> create_axiom('%subject_id SubClassOf: %object_id', "direct:metadata");
```
Presumably you want those axioms to be created for both Uberon and CL classes?

Then you could either do:

```
(object==UBERON:* || object==CL:*) && predicate==semapv:crossSpeciesExactMatch -> ...;
```

But, since you already dropped all mappings with an object that is not in Uberon or CL above, you no longer need to filter on object and you could simply do:

```
predicate==semapv:crossSpeciesExactMatch -> ...;
```

matentzn replied:
Awesome thank you!

gouttegd commented on src/ontology/config/mappings-to-uberon-bridge.rules:
```diff
@@ -0,0 +1,16 @@
+prefix CL:    <http://purl.obolibrary.org/obo/CL_>
+
+# Make sure UBERON/CL classes, and only UBERON/CL classes, are on the object side
+subject==UBERON:* -> invert();
```
As explained in another comment:

```
(subject==UBERON:* || subject==CL:*) -> invert();
!(object==UBERON:* || object==CL:*) -> stop;
```

matentzn replied:
PErfect thanks!

gouttegd commented on src/ontology/phenio-odk.yaml:
```diff
@@ -11,129 +12,206 @@ export_formats:
+      mirror_from: http://purl.obolibrary.org/obo/uberon/uberon.sssom.tsv
+    - id: cl
+      maintenance: mirror
+      mirror_from: https://raw.githubusercontent.com/obophenotype/uberon/refs/heads/master/src/mappings/cl.sssom.tsv
```
Err, why not getting the CL set from its “official” PURL? `http://purl.obolibrary.org/obo/cl/cl.sssom.tsv` ?

matentzn replied:
Fixed, I thought I tried to find it but I didn't now I see. :)

gouttegd commented on src/ontology/phenio-odk.yaml:
```diff
@@ -11,129 +12,206 @@ export_formats:
+        - http://purl.obolibrary.org/obo/FBcv_
+    
+    # Species specific anatomy Ontologies
+    # Consider replacing by Uberon metazoan?
```
If you mean “composite-metazoan”, keep in mind what I told you about the “unfolding” that happens in that product. It may _not_ be what you need, and so that comment may be misleading for any future reader.

matentzn replied:
I have fixed that now, thaks!

gouttegd commented on src/ontology/phenio.Makefile:
```diff
@@ -17,24 +16,72 @@ RELEASE_ASSETS_AFTER_RELEASE=$(foreach n,$(RELEASE_ASSETS), ./$(n))
+		remove --select "<http://purl.obolibrary.org/obo/OPL_*>" \
+		remove --select "<https://bioregistry.io/lipidmaps*>" \
+		remove --term owl:Thing --term owl:Nothing \
+		rename --mappings config/property-map.tsv --allow-missing-entities true --allow-duplicates true \
```
Not wanting to push too hard for my own software, but if you want to use a real SSSOM file for the “property map”, you could use the `sssom:rename` command of the SSSOM plugin.

That’s [what we do](https://github.com/obophenotype/uberon/blob/5b006f5af53983b4f2e03564a429c7e3e496accb/src/ontology/uberon.Makefile#L389) in Uberon, using [this mapping set](https://github.com/obophenotype/uberon/blob/master/src/mappings/import-corrections.sssom.tsv) as source.

matentzn committed:
Add proper NCBIGENE and HGNC imports

Previously we were using the Mondo gene imports as a basis, which led to a bunch of dangling gene classes. This should now be fixed.

matentzn committed:
Fix bridges in PHENIO

This is a big refactor, removing all the uberon and upheno bridges as hard imports and replacing them by SSSOM based pipelines using java-sssom.

caufieldjh commented:
What remains to be done on this PR before it's ready to merge? Just reviewing?

matentzn commented:
@caufieldjh It is ready from my perspective!

caufieldjh commented:
Great - thanks @matentzn and @gouttegd !
